File: week2/C++/gcd.cpp
#include <iostream>

long long gcd_optimized(long long a, long long b) {
  if (b == 0)
    return a;
  else{
    if (a >= b)
      return gcd_optimized(b, a%b);
    else
      return gcd_optimized(a, b%a);
  }
}

int main() {
  long long a, b;
  std::cin >> a >> b;
  std::cout << gcd_optimized(a, b) << std::endl;
  return 0;
}

File: week2/python/gcd.py
# Uses python3

def gcd_optimized(a, b):
    if b == 0:
    	return a
    else:
    	if a >= b:
    		return gcd_optimized(b, a%b)
    	else:
    		return gcd_optimized(a, b%a)


if __name__ == "__main__":
    a, b = map(int, input().split())
    print(gcd_optimized(a, b))
